Allow null option for properties

andreashug opened the pull request:
Proposal for solving #47

The Swagger specification in version 2 doesn't support 'null' as
property value. This adds a vendor extension 'x-nullable' which
allows properties to be 'null' if set to 'true'.

msander commented:
+1

analogue commented:
Thanks for the PR. Haven't had time to review yet so it will probably be next week.

analogue commented on tests/swagger20_validator/type_validator_test.py:
```diff
@@ -34,3 +34,25 @@ def test_validate_when_not_a_parameter_schema(m_draft4_type_validator,
+@patch('jsonschema._validators.type_draft4')
+def test_skip_when_nullable_property_schema_and_value_is_None(
+        m_draft4_type_validator, minimal_swagger_spec):
+    param_schema = {'name': 'foo', 'x-nullable': True, 'type': 'string'}
```
I think you mean for this to be `prop_schema` with no `name` key.

andreashug replied:
You are right. I'll change that.

analogue commented on tests/swagger20_validator/type_validator_test.py:
```diff
@@ -34,3 +34,25 @@ def test_validate_when_not_a_parameter_schema(m_draft4_type_validator,
+@patch('jsonschema._validators.type_draft4')
+def test_validate_when_not_nullable_property_schema_and_value_is_None(
+        m_draft4_type_validator, minimal_swagger_spec):
+    param_schema = {'name': 'foo', 'x-nullable': False, 'type': 'string'}
```
Same here

andreashug replied:
Same here ;-)

analogue commented:
What is your interpretation of `x-nullable` when `required==True`?  If I'm reading this correctly,  when a property is nullable and required and the value passed is None, then that makes it OK.  Hence, there would have to be changes to `unmarshal_primitive(..)` and `marshal_primitive(..)`

andreashug commented:
My understanding of the spec is that `required==True` only means the property must exist and `x-nullable==True` allows `null` as value.
 
 | `required==False`  |  `required==True` |
----- | ----- | ------
`x-nullable==False` | `{ }->ok` `{"x": null}->fail`  |  `{ }->fail` `{"x": null}->fail`
`x-nullable==True` | `{ }->ok` `{"x": null}->ok` | `{ }->fail` `{"x": null}->ok`

I'm not sure if this is possible because `unmarshal_object(..)` re-introduces missing properties with `None`. What do you think?

andreashug commented:
While searching for a good solution, I discovered that JSON Schema allows a list of types - e.g. `
{
    "schema": {
        "type": ["string", "null"]
    }
`
Actually I think this would be a much better way to solve my problem and it works well with `jsonschema`. Unfortunately it's not supported bravado-core yet. I'll try to come up with a proposal so this pull request can be postponed.

wasbazi commented:
This would be a really nice feature to have! There is a long thread on the OpenAPI-Specification (formerly The Swagger Specification) repository discussing how they want to solve this and leaning towards the `x-nullable` case https://github.com/OAI/OpenAPI-Specification/issues/229

Unfortunately it seems from the discussion that using arrays for type is not intended with the 2.0 spec (which seems to be indicated as feature frozen) https://github.com/OAI/OpenAPI-Specification/issues/229#issuecomment-164587628

Are there things I can do to help get this PR finished?

cyncyncyn commented:
As @wasbazi says, this would be a nice feature to have!
I also offer my help in order to get this PR merged.
Thanks!

sjaensch commented:
The CR has an outstanding issue, see https://github.com/Yelp/bravado-core/pull/64#issuecomment-175177203
The proposed solution of going with a list of types as allowed by JSON Schema will not be possible, the Swagger spec authors already clarified that Swagger 2.0 is not supposed to allow a list of types.

normalex commented:
Can we merge this? people are eager to see it released.

andreashug commented:
Finally found some time to continue with this.
1) Rebased on recent master
2) Fixed test according to https://github.com/Yelp/bravado-core/pull/64#issuecomment-170170418
3) Integration tests for unmarshaling

The good news is that unmarshaling doesn't seem to be affected (see matrix in https://github.com/andreashug/bravado-core/blob/c581aaba8924afe340f124375015afc9f5743762/tests/unmarshal/unmarshal_nullable_test.py)

The not so good news however is, I'm not sure how marshaling should behave in combination with `x-nullable`. Currently, `marshal_object()` drops keys whose value is `None`, which leads to a ValidationError (required property) -> https://github.com/Yelp/bravado-core/blob/7d13d2512f85983eca086db1eb0ddab797b22499/bravado_core/marshal.py#L129-L130. One might think, it should keep the key if `x-nullable` is `True`. Or we could say that `x-nullable` is only for properties and not accounted in parameter specs.

Any opinions on that?

sjaensch commented:
Great stuff @andreashug! I'm still travelling, I'll take a closer look at the patch on Tuesday.
Just to make sure I understand the issue you're seeing: If an endpoint expects a parameter of type object that has a `x-nullable` field then we get a ValidationError?

sjaensch commented on tests/swagger20_validator/type_validator_test.py:
```diff
@@ -34,3 +34,25 @@ def test_validate_when_not_a_parameter_schema(m_draft4_type_validator,
+        m_draft4_type_validator, minimal_swagger_spec):
+    prop_schema = {'x-nullable': False, 'type': 'string'}
+    args = (None, prop_schema['type'], None, prop_schema)
+    list(type_validator(minimal_swagger_spec, *args))
```
I'd use keyword arguments here as well, just like you did in the test above. It's more readable.

andreashug replied:
This is because `assert_called_once_with()` expects the arguments without keywords. The existing tests in this file do it this way so I aligned with them ;-)

sjaensch commented on tests/unmarshal/unmarshal_nullable_test.py:
```diff
@@ -0,0 +1,87 @@
+    except ValidationError as e:
+        result = e
+
+    if nullable is False and not value:
```
`not value is None` is more correct here. Which reminds me, testing with an empty string (or a bool with the value False) might not be a bad idea.

You could split out the case that is supposed to fail into a separate test function so that you don't have to have this logic here at all, but I'm not sure if that actually improves things.

sjaensch commented on tests/unmarshal/unmarshal_nullable_test.py:
```diff
@@ -0,0 +1,87 @@
+    elif nullable is False and value['x'] is None:
+        assert result.message == "None is not of type 'string'"
+    else:
+        result == value
```
Having such a long `if/elif` when doing test result assertion is a good indicator that your test tries to do too many things at once. In the failure case you could then use `pytest.raises` to check that the ValidationError was thrown and use mark.parametrize for the expected error message.

In such a case where the expected outcome is dependent on several parameters you could use the singular version of mark.parametrize (i.e. only one decorator that specifies all parameter combinations explicitly) together with the expected outcome.

andreashug replied:
You are totally right. When I started to write these integration tests, I actually didn't think that they should go into repo (still not sure about that) because they only clarify what happens in which scenario. But anyway, here's the refactoring =)

andreashug commented:
Regarding `marshal_object()`: The `ValidationError` is raised when `x-allownull` is `True`, the value is required the value is set to `None`. See https://dpaste.de/D0ax

sjaensch commented:
This is great! Could you please create another paste? I didn't manage to open it in time. 😁

andreashug commented:
Sure: http://pastebin.com/U3xeA48R  (Seems like there is a bug in expiration settings on dpaste.de)

sjaensch commented:
Thanks again @andreashug! While I think we could even ship this without fixing the edge case about a required x-nullable parameter, I think the fix is not that difficult:

```
diff --git a/bravado_core/marshal.py b/bravado_core/marshal.py
index 536e627..8a3771a 100644
--- a/bravado_core/marshal.py
+++ b/bravado_core/marshal.py
@@ -125,13 +125,14 @@ def marshal_object(swagger_spec, object_spec, object_value):
     result = {}
     for k, v in iteritems(object_value):

-        # Values cannot be None - skip them entirely!
-        if v is None:
-            continue
-
         prop_spec = get_spec_for_prop(
             swagger_spec, deref(object_spec), object_value, k)

+        # Values cannot be None unless x-nullable is set
+        is_nullable = prop_spec and prop_spec.get('x-nullable', False)
+        if v is None and not is_nullable:
+            continue
+
```

With it the test you wrote (which is in your dpaste) passes. We might want to handle the other cases (nullable primitives etc.) as well, but honestly I think it's not that important, since the use case for x-nullable are model properties in responses. You're welcome to incorporate this small change as well as the test in the branch though.

I'm going to find some time to test this with bravado and make sure it does what we want. 😊

sjaensch commented:
An manual test using bravado to query an endpoint that returns a model with an x-nullable field worked perfectly! I'm going to merge this in and work on a second pull request that will contain my above change, a version bump and a big thank you to @andreashug in the changelog. 😊

andreashug commented:
Yay, thanks  \o/
